ch4/rma: optimize flushes with recv AM progress interval

minsii opened the pull request:
Every RMA flush always polls AM progress at least once to handle
received AM. The progress polling causes unnecessary overhead when all
RMA are native. This patch defines progress interval CVARs to reduce
such overhead. The recv AM progress will be polled with long interval at
beginning. Polling becomes active only when an incoming AM has been
received on the local process.

Use MPIR_CVAR_CH4_RMA_RECV_AM_PROGRESS_INTERVAL to make explicit control
or enable the above runtime optimization.

Use MPIR_CVAR_CH4_RMA_RECV_AM_PROGRESS_INTERVAL_LONG to set the value
of long interval when no recv AM arrives.

This feature disabled by default. Set
MPIR_CVAR_CH4_RMA_RECV_AM_PROGRESS_INTERVAL to -1 to enable.

## Pull Request Description

<!--
Insert description of the work in this merge request (above this comment),
particularly focused on _why_ the work is necessary, not _what_ you did.
-->

<!-- AUTHOR: After creating this merge request, check off each of the following items as you complete them. -->

## Expected Impact

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] You or your company has a signed contributor's agreement on file with Argonne
* [ ] For non-Argonne authors, request an explicit comment from your companies PR approval manager

minsii commented:
Alternative solution to replace #4826

minsii commented:
test:mpich/ch4/most

hzhou commented on src/mpid/ch4/src/ch4r_win.h:
```diff
@@ -25,6 +25,36 @@
+        across all windows.
+        A value of 1 means always enable receive AM progress polling in every flush call.
+        A value of 0 means completely disable explicit progress polling for receive AM.
+        A value of -1 means the interval is automatically optimized at runtime.
```
Can we make it an enum or boolean?

minsii replied:
The value may be an arbitrarily large integer as the interval. I am afraid mixing types will cause confusion.

hzhou replied:
I think we probably want to rename this CVAR so it is clear as an opt-in switch. We are essentially removing the "high frequency" control as we discussed, right? I think trying to mixing the semantics of that control to the of/off switch control is confusing.

hzhou replied:
Just for understanding, what is the difference between value `1` and `0`?

minsii replied:
> Just for understanding, what is the difference between value `1` and `0`?
> 
> It seems 0 means never poll AM progress in `MPI_Win_flush`, right? I don't think we need that option.

It is for developers/users who want to test with AM progress completely disabled. I think CVAR can be used for this purpose.

hzhou replied:
Currently the model in my mind needs following two CVARs:
`MPIR_CVAR_CH4_RMA_ENABLE_SKIPPING_AM_PROGRESS`, and
`MPIR_CVAR_CH4_RMA_DEFAULT_AM_PROGRESS_INTERVAL`.

minsii replied:
> I think we probably want to rename this CVAR so it is clear as an opt-in switch. We are essentially removing the "high frequency" control as we discussed, right? I think trying to mix the semantics of that control to the on/off switch control is confusing.

Do you have any better CVAR definition in mind?

minsii replied:
Having a separate ENABLE CVAR is fine to me. But your suggested CVARs do not allow me to statically set the interval no matter whether incoming AM has been received. This functionality is an valid option for many users.

hzhou replied:
> But your suggested CVARs do not allow me to statically set the interval no matter whether incoming AM has been received. This functionality is an valid option for many users.

The case is we know there is AM atomic messages, and we don't want to poll AM progress always? Wouldn't that guarantee the extra latency?

I assume you are not describing the case when there is AM atomic messages from *other* window, since I thought that is the case you want simplify away for now.

minsii replied:
"statically set the interval no matter whether incoming AM has been received" allows the user to tune the interval no matter the existence of recv_am. OFI has similar feature too, to allow user to tune the interval for waking up the async thread. I don't expect an end user to use it. But it is an useful thing for researchers or advanced users who wants to tune async progress v.s. latency.

> I assume you are not describing the case when there is AM atomic messages from other window, since I thought that is the case you want simplify away for now.
It is irrelevant to the above case.

hzhou commented on src/mpid/ch4/src/ch4r_win.h:
```diff
@@ -559,8 +589,10 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_mpi_win_flush(int rank, MPIR_Win * win)
+    int poll_once = MPIDIG_rma_recv_am_always_poll()? 1 : 0;
+    MPIDIU_PROGRESS_WHILE((target_ptr && (MPIR_cc_get(target_ptr->remote_cmpl_cnts) != 0 ||
+                                          MPIR_cc_get(target_ptr->remote_acc_cmpl_cnts) != 0)) ||
+                          poll_once-- > 0);
```
clever :)

hzhou commented on src/mpid/ch4/src/ch4_impl.h:
```diff
@@ -765,70 +765,71 @@ MPL_STATIC_INLINE_PREFIX void MPIDIG_win_remote_cmpl_cnt_decr(MPIR_Win * win, in
             break;
         }
     }
+    return allcompleted;
```
Yes! I wanted do this for a while. Actually I think I did it in one of my other PR (don't remember which).

hzhou commented on src/mpid/ch4/src/ch4r_win.h:
```diff
@@ -177,9 +177,8 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_mpi_win_complete(MPIR_Win * win)
-        MPIDIU_PROGRESS();
-    } while (!MPIDIG_win_check_group_local_completed(win, ranks_in_win_grp, group->size));
+    MPIDIU_PROGRESS_DO_WHILE(!MPIDIG_win_check_group_local_completed(win, ranks_in_win_grp,
+                                                                     group->size));
```
👍

minsii commented on src/mpid/ch4/generic/am/mpidig_am.h:
```diff
@@ -77,6 +77,10 @@ typedef int (*MPIDIG_am_target_msg_cb) (int handler_id, void *am_hdr,
     MPIDIG_am_origin_cb origin_cbs[MPIDI_AM_HANDLERS_MAX];
+    /* control parameters for global progress of RMA receive active messages */
+    MPL_atomic_int_t rma_recv_am_active;        /* Indicates whether any incoming RMA AM has been received.
+                                                 * Set inside each target callback.*/
```
Have the same concern. Same for `rma_recv_am_poll_interval`. But semantically I don't think it makes sense to set recv_am progress per window. Because everytime we make progress for it, it actually handles am globally. So in my opinion. recv_am is always global. We have discussed this yesterday.

We can study the potential performance loss as next step. If we find critical issue, we can fix from there.

minsii replied:
Will add a node in code for the performance concern.

minsii commented on src/mpid/ch4/generic/am/mpidig_am.h:
```diff
@@ -77,6 +77,10 @@ typedef int (*MPIDIG_am_target_msg_cb) (int handler_id, void *am_hdr,
+    /* control parameters for global progress of RMA receive active messages */
+    MPL_atomic_int_t rma_recv_am_active;        /* Indicates whether any incoming RMA AM has been received.
+                                                 * Set inside each target callback.*/
+    MPL_atomic_int_t rma_recv_am_poll_interval;
```
`rma_recv_am_poll_interval` only counts the number of called flushes.

hzhou replied:
You mean this is a `counter`, not an `interval`?

minsii replied:
I see. So will `rma_recv_am_poll_cntr` make it clear? Will change.

minsii commented on src/mpid/ch4/src/ch4_impl.h:
```diff
@@ -1146,4 +1146,38 @@ MPL_STATIC_INLINE_PREFIX MPI_Op MPIDIU_win_acc_get_op(int index)
+        /* User explicitly disables polling */
+        always_poll = false;
+    }
+
```
It is not completely disable polling in flushes. We still poll with certain interval

hzhou replied:
I assume if this returns true and the INTERVAL CVAR is set to `0`, then the polling is essentially disabled, right?

minsii replied:
Returning false will only disable poll for the calling flush. It needs to internally check whether the interval is met. If so, it needs to return true -- meaning the calling flush will poll progress.

raffenet replied:
Let's back up a second. I thought that the goal was to completely disable AM progress when the application (OSHMPI) doesn't need it, right? And right now, AM progress is called during every progress loop iteration? What don't we just add a CVAR to disable it and avoid this interval discussion for right now? Or is the interval-based polling also a critical piece for this release?

I remember a discussion about calling internal abort if an active message is issued on a window that expects only native operations. That sounds OK to me, too.

raffenet replied:
Oh, I guess the problem is the CVAR is global and not per-window...

hzhou replied:
The interval is needed for correctness. If we never poll AM, even the given process is entirely doing native RMA, the other process may still issue am rma targeting the given process, and it will stall due to lack of progress.

minsii replied:
Right. Without extra hints, we cannot know whether any remote process may issue AM-based RMA to the given process. With interval, it allows the given process to poll progress with a long interval (or low frequency) at beginning, and change to a shorter interval (more frequent, essentially interval=1) polling once it sees any incoming AM-based RMA. 

This way will allow OSHMPI to skip the unnecessary polling for incoming RMA AM in most fast cases (i.e., contig put/get, atomics), while still making progress when an unusual strided put/get (AM-based PUT/GET in MPI) is issued.

minsii commented on src/mpid/ch4/src/ch4_impl.h:
```diff
@@ -1146,4 +1146,38 @@ MPL_STATIC_INLINE_PREFIX MPI_Op MPIDIU_win_acc_get_op(int index)
+
+MPL_STATIC_INLINE_PREFIX void MPIDIG_rma_recv_am_activate(void)
+{
+    MPL_atomic_store_int(&MPIDIG_global.rma_recv_am_active, 1);
```
It needs to be `rma_recv_am`, to be distinguished from the outgoing am for RMA. We poll this progress only from incoming recv_am.

hzhou replied:
If you found incoming am, why don't you want to always poll progress? Polling am progress is for getting incoming am, right?

minsii replied:
I do get your above comment. (1)incoming am == recv_am, (2)outgoing am == locally issued am. This PR is to optimize progress for (1) -- we control the progress internal before finding any (1), right?

minsii commented on src/mpid/ch4/src/ch4r_rma_target_callbacks.c:
```diff
@@ -1122,6 +1122,7 @@ int MPIDIG_put_ack_target_msg_cb(int handler_id, void *am_hdr, void *data, MPI_A
     if (is_async)
         *req = NULL;
 
+    MPIDIG_rma_recv_am_activate();
```
I don't want to make the code too complex. An atomic write should not cause visible overhead in the AM path, in my opinion. But we can certainly study it. For now, I don't want to make the code too complex. And semantically, I think recv_am is global not per-window.

hzhou commented on src/mpid/ch4/src/ch4_impl.h:
```diff
@@ -1146,4 +1147,40 @@ MPL_STATIC_INLINE_PREFIX MPI_Op MPIDIU_win_acc_get_op(int index)
+        } else {
+            always_poll =
+                ((interval + 1) % MPIR_CVAR_CH4_RMA_RECV_AM_PROGRESS_INTERVAL_LONG ==
+                 0) ? true : false;
```
@minsii This branch clearly says the semantics is not *always*, so I would suggest remove the "always" (in addition of the "recv") from the function name, simply call it `MPIDIG_rma_need_poll_am()`.

minsii replied:
Agree to remove `always`. But I do want to keep the meaning that this function defines only the progress demand for incoming RMA AM. E.g., even this function returns false, we still need check outgoing RMA AM in flush calls and decide to poll progress if it found any outgoing RMA AM.

hzhou replied:
If you really insist, how about `MPIDIG_rma_need_poll_am_recv()`? I don't really like having the extra `recv` suffix, but since I got the `need_poll_am` part, I think I can tolerate the suffix. @raffenet What do you think?

minsii replied:
You made the point that `am_recv` or `recv_am` might be unclear as it also includes receiving the ack on the origin process (which we don't care here). I agree with the point. So, how about `MPIDIG_rma_target_need_poll_am` or `MPIDIG_rma_need_poll_target_am`?

hzhou replied:
>  how about `MPIDIG_rma_target_need_poll_am` or `MPIDIG_rma_need_poll_target_am`?

Well, when we read this name, the natural question is -- "which target"? That's why I think the word "target" on its own is very confusing.

minsii replied:
Target is the standard term to describe the process who receives RMA operation in a RMA communication.

hzhou replied:
The confusing part is not what "target" means in the context of RMA. The confusing part is target didn't say anything about which origin, which target, or which process. There are plenty of variables names having "target" but often means other processes. Here you want it to mean the opposite, so it is confusing.

hzhou replied:
How about "non-origin"?

hzhou replied:
Back up a little, the full sentence your wanted to say here is `MPIDIG_rma_need_poll_am_because_of_[this_reason]` and we are discussing a good word to describe this reason. But even we have that word, if your take out the "because_of" part, the phrase is very confusing: `MPIDIG_rma_need_poll_am_reason`? 

I guess I am back to insist on just `MPIDIG_rma_need_poll_am` since that is complete on its own.

minsii replied:
We specifically use `origin_cbs` and `target_msg_cbs` to distinguish the AM received on origin process (ACK) and that received on target process in our code. The new flag is set only when any `target_msg_cbs` is called, and the function decides the polling strategy only for progressing `target_msg_cbs`. Thus, I need have the meaning in the name.

I think we cannot make conclusion for the naming. Thus, hope @raffenet can raise the third voice so we can finish the PR.

raffenet replied:
Of the names suggested in this thread, I prefer `MPIDIG_rma_need_poll_am`. I don't think the reason needs to be stated in the API. The function is there to let us know if we need to poll (progress?) active messages for RMA. Why we need to poll is an implementation detail.

minsii replied:
Alright. Will make changes based on the naming of `MPIDIG_rma_need_poll_am`, although I still think it is confusing as we need check additional conditions outside of the function.

hzhou replied:
> I still think it is confusing as we need check additional conditions outside of the function.

I see. If you don't like it because the function is incomplete, then why don't you roll the other condition into this function as well?

minsii replied:
We need check local completion for flush_local but remote completion for flush. I don't like mix the two cases in a single function and pass a parameter to distinguish them. IMO, they should belong to the original functions.

hzhou replied:
Do we need poll am in `flush_local` when there is no outstanding local operations?

minsii replied:
I don't see why we need skip poll am in `flush_local`. We want to be consistent for all flush calls.

hzhou replied:
The purpose of MPI_Win_flush_local is to be able to reuse the local buffer. So semantically user won't expect this function to help any remote progress.

It is not a strong reason, since MPI_Win_flush doesn't guarantee remote progress either (although some user may expect that). But it is the same reason that we decide not to call general progress in, for example, `MPI_Put`. I think narrow down our scope to `MPI_Win_flush` is good for now.

minsii replied:
User would expect some level of progress in flush_local. E.g., such loop may exist in user program and low-frequency progress is good to have.
```while(a<100) {fop(a/*result buf*/); flush_local(); /*complete atomic get*/}```

hzhou replied:
The MPI standard only mentions `put`, `get`, and `accumulate` functions for `MPI_Win_flush_local`, though.

But as I relented, there isn't a strong reason that we choose to poke additional am progress in `MPI_Win_flush` vs poking progress in any other MPI functions such as `MPI_Put`. In a passive window, there isn't any semantics to guarantee progress. We choose `MPI_Win_flush` simply as a compromise, because that is a function that user would expect to be most expensive (other than those epoch closing syncs, which we poll am progress always). If you want to include `MPI_Win_flush_local`, it is your call. Then of course, your code and semantics gets a bit more complex. I'll leave to your judgement.

minsii replied:
Please trust me that many places in the standard mention `accumulate`  which actually means all accumulate-like operations, including acc, get_acc, fop, cswap. (I know it is confusing).

I still think it is straightforward and reasonable to also add the progress opt in `flush_local`. Dropping it will not make the semantics of `MPIDIG_rma_need_poll_am` more clear. Also, moving the origin-issued-am check into `MPIDIG_rma_need_poll_am` only for `flush_{all}` will make the code inconsistent. 

Anyway, I will rename the function/flag/CVAR, but let's leave the code logic as is.

hzhou commented on src/mpid/ch4/src/ch4_impl.h:
```diff
@@ -1146,4 +1147,40 @@ MPL_STATIC_INLINE_PREFIX MPI_Op MPIDIU_win_acc_get_op(int index)
+    return always_poll;
+}
+
+MPL_STATIC_INLINE_PREFIX void MPIDIG_rma_recv_am_activate(void)
```
This name is another sticking point. It is not "activate" "recv_am", right? It is merely updating/setting a flag indicating we have active messages targeting this process. So I would like to see the semantics of "updating/setting flag" intuitive from the name.

minsii replied:
I tried to follow the activate term used to enable/disable progress hooks. But yes, we don't really activate recv_am. How about `MPIDIG_rma_set_target_am_arrived`?

hzhou replied:
I like the `MPIDIG_rma_set_` part. "target_am" is very confusing. "arrived" is also not exactly the intention. The intention is we don't care about "arrived" am message, but having received am message indicates that we may receive future unexpected am messages. It is too bad you don't like `unexp`, I think it really describes exactly your intention. @raffenet what is your opinion?

raffenet replied:
Similar to the other case, I don't think we need to state "why" we're updating the am flag in the API. That is determined at the call site. Would this API be used if the origin process issues its own active message RMA? Or are we explicitly distinguishing when active messages have been issued to this process (target)?

minsii replied:
My intent is to explicitly distinguish when active messages have been issued to this process (target). Because we never reduce the polling frequency on the origin process when the origin process issues its own active message RMA.

minsii commented:
test:mpich/ch4/most

minsii commented:
All fixes should have been addressed. Please check again. The last commit is only for testing purpose.
test:mpich/ch4/most

raffenet reviewed (approved):
This looks OK to me. The unconditional atomic store in each callback seems a little heavy-handed, but I don't have a better solution right now. We could check the value before setting it, but that would require a read+branch. Unconditional store is probably just as good. I guess we can leave it for future study, as noted.

hzhou commented:
test:mpich/custom
device: ch4:ofi
env: MPIR_CVAR_CH4_RMA_ENABLE_DYNAMIC_AM_PROGRESS=1

hzhou reviewed (approved):
LGTM. Running a custom test just to confirm.

Could you update the PR description to reflect the CVAR change?

minsii commented:
Updated PR description. Removed the last commit for Jenkins test.

test:mpich/ch4/most
test:mpich/custom
device: ch4:ofi
env: MPIR_CVAR_CH4_RMA_ENABLE_DYNAMIC_AM_PROGRESS=1

raffenet commented:
@minsii ok to merge?
